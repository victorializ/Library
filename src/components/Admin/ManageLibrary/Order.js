import React from 'react';
import { Card, Button } from 'semantic-ui-react';

import { Order as OrderTemplate } from '../../Order/Order';
import { constants } from '../../../i18n';

const Order = ({ item, finishOrder }) => 
    <Card>
        <Card.Content>
            <OrderTemplate key={item.bookingId} item={item} />
        </Card.Content>
        <Card.Content extra>
            <div>
                {`${constants.customer}: ${item.user.email}`}
            </div>
            <Button fluid onClick={() => finishOrder(item.bookingId)}>
                {constants.finishOrder}
            </Button>
        </Card.Content>
    </Card>

export { Order };