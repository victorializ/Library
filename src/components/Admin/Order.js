import React, { useState } from 'react';
import { Button, Dropdown } from 'semantic-ui-react';

import { constants } from '../../i18n';
import { LoadingComponent } from '../hocs/LoadingComponent';
import { ErrorMessage } from '../UITable/ErrorMessage';
import { useRequest } from '../../hooks';
import { getAllBooks, getAllUsers, order as orderRequest } from '../../services/http-client';

function Order() {
    const allUsersResponse = useRequest(getAllUsers);
    const allBooksResponse = useRequest(getAllBooks);

    const [ orderProcess, setOrderProcess ] = useState(false);
    const [selectedUser, setSelectedUser] = useState(null);
    const [selectedBook, setSelectedBook] = useState(null);
    const response = useRequest(orderRequest, [selectedUser, selectedBook], orderProcess);
  
    const getDropdownOptions = (array, mapper) => {
        return array === null ? 
        [] : array.map(item => mapper(item));
    }

    const bookToDropdownOption = book => {
        return {
            key: book.bookId, 
            value: book.bookId,
            text: book.name
        };
    };

    const userToDropdownOption = user => {
        return {
            key: user.userId,
            value: user.userId, 
            text: `${user.firstName} ${user.lastName} (${user.email})`
        };
    };

    return (
      <div className="admin__order">
        {
          orderProcess &&  <LoadingComponent {...response} />
        }
        <div className="admin__order__items">
          {
            allBooksResponse.error.message ?
              <ErrorMessage text={allBooksResponse.error.message} /> 
                :
              <Dropdown 
                search
                selection
                loading={allBooksResponse.loading} 
                onChange={(e, {value}) => {
                    setOrderProcess(false);
                    setSelectedBook(value);
                }}
                value={selectedBook}
                placeholder={constants.books}
                options={getDropdownOptions(allBooksResponse.data, bookToDropdownOption)}
              />
          }
          {
            allUsersResponse.error.message ?
              <ErrorMessage text={allUsersResponse.error.message} /> 
                :
              <Dropdown 
                search
                selection
                loading={allUsersResponse.loading} 
                onChange={(e, {value}) => {
                    setOrderProcess(false);
                    setSelectedUser(value);
                }}
                value={selectedUser}
                placeholder={constants.usersList}
                options={getDropdownOptions(allUsersResponse.data, userToDropdownOption)}
              />
          }
        </div>
        <Button 
          className='registration__button'
          onClick={() => setOrderProcess(true)}
          disabled={!selectedUser || !selectedBook}>
          {constants.order}
        </Button>
      </div>
    );
  }

export { Order };