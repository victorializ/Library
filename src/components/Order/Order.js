import React from 'react';
import { Card } from 'semantic-ui-react';
import moment from 'moment';

import { constants  } from '../../i18n';
import { colors } from '../../assets/semantic-colors';
import { formatData } from '../../services/format-data/order';

function Order({item}) {
    const {
        name, 
        bookYear,
        date, 
        isFinished
    } = formatData(item);

    const statuses = {
        finished: {
            message: constants.returned, 
            color: colors.grey
        }, 
        active: {
            message: `${constants.return} ${moment(date).calendar()}`, 
            color: colors.primary
        }, 
        expired: {
            message: constants.expired, 
            color: colors.red
        }
    };
    
    const isReturnDateAfterToday = () => {
        const now = moment();
        const returnDate = moment(date);
        return returnDate.diff(now) > 0;
    }
    const status = () => 
        isFinished ? 
            statuses.finished 
                :
            isReturnDateAfterToday() ?
                statuses.active
                    :
                statuses.expired;
   
    return (
        <Card className='book' color={status().color}>
            <Card.Content>
                <Card.Header>
                    {name}
                </Card.Header>
                <Card.Meta>
                    {bookYear}
                </Card.Meta>
                <Card.Description>
                    {`${constants.status}: ${status().message}`}
                </Card.Description>
            </Card.Content>
        </Card>
    );
}

export { Order };