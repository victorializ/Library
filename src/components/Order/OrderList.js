import React  from 'react';

import { Order } from './Order';
import { ItemsList } from '../ItemsList';

const OrdersList = ({data}) => <ItemsList data={data} Component={Order} />

export { OrdersList };