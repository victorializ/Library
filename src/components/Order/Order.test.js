import React from 'react';
import { shallow } from 'enzyme';

import { Order } from './Order';

describe('Order component', () => {
    it('renders correctly', () => {
        expect(shallow(<Order item={{book: {}}}/>)).toMatchSnapshot();
    }); 
});