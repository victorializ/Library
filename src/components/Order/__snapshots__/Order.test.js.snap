// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order component renders correctly 1`] = `
<Card
  className="order__book"
  color="red"
>
  <CardContent>
    <CardHeader />
    <CardMeta />
    <CardDescription>
      Status
      : 
      Expired!
    </CardDescription>
  </CardContent>
</Card>
`;
